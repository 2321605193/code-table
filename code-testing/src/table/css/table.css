.common-padding {
    padding: 5px 20px
}

.fj-table {
    display: inline-block;
    margin: auto;
    margin-top: 20px;
    margin-left: 100px;
}
.fj-table-content {
    display: inline-block;

    
    border: 1px solid #f0f0f0;
}

.fj-table-content table{
    display: inline-block;
    width: 100%;
    border-collapse: collapse;
 }
 

 .fj-table-content .fj-table-body tbody {
    display: block;
    height: 420px;
    overflow-y: scroll;
 }

.fj-table__row {

    border-bottom: 1px solid #f0f0f0;
}

/* table thead, table tbody ,table thead tr, table tbody tr {
    display: table;
     table-layout: fixed;
} */

.fj-table__header-cell {
    display: flex;
    justify-content: center;
    align-items: center;
    min-height: 45px;
}


.sort-button {
    display: inline-block;
    width: 16px;
    margin-left: 12px;
}

.sort-button__icon--default { 
    transform: rotate(0deg);
    fill: currentColor;
    color: rgba(194, 194, 194, 1)
}



.sort-button__icon--asc { 
    transform: rotate(180deg);
    fill: currentColor;
    color: #18a058;
}

.sort-button__icon--desc {
    transform: rotate(0deg);
    fill: currentColor;
    color: #18a058;
}

/* 设置滚动条的样式 */
::-webkit-scrollbar {
    width: 4px;
}
/* 滚动槽 */
::-webkit-scrollbar-track {
    -webkit-box-shadow:inset006pxrgba(0,0,0,0.3);
    border-radius:10px;
}
/* 滚动条滑块 */
::-webkit-scrollbar-thumb {
    border-radius:10px;
    background:rgba(0,0,0,0.1);
    -webkit-box-shadow: inset006pxrgba(0,0,0,0.5);
}
::-webkit-scrollbar-thumb:window-inactive {
    background:rgba(39, 38, 38, 0.4);
}




.pagination-content {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    width: 100%;
}