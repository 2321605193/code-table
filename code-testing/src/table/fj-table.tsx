
import { defineComponent, reactive, ref, provide, computed, toRefs } from '@vue/composition-api'
import { tableProps, PaginationOptions } from './types'
import Table from './table';
import Pagination from './pagination';



export default defineComponent({
  name: 'fjTable',
  props: tableProps,
  setup(props) {

    
    let tableData = reactive(props.data)
    let columns = reactive(props.columns)
    const showHeader = ref(props.showHeader)
    let showPagination = ref(props.showPagination)


    let { paginationOptionsChange, paginationOptions } =  usePagination();



    // 获取table总数据
    let getTableColumns = computed(() => {
      return columns
    })

    let total = computed(()=> {
      return tableData.length;
    })

    provide('tableProvide', {
      total,
      filterTableData: computed(() => {
        let templateData = tableData.slice();
        if (getSortRule.value) {
          templateData.sort((curr, next)=> getSortRule.value(curr, next))
        }
        if (showPagination.value) {
          return templateData.slice((paginationOptions.value.page - 1 ) * paginationOptions.value.size, paginationOptions.value.page * paginationOptions.value.size);
        }
        return templateData
        
      }),
      getTableColumns,
      showHeader,
      paginationOptionsChange,
      paginationOptions,
      changeSortRuleAndSortKey,
      getSortKey
    })

    return () => {
      let pagination = showPagination.value ? Pagination : null;

      return (
        <section class='fj-table'>
          <Table /> 
          <br/>
          <pagination />
        </section>
      )
    }
  },
})


// 分页器相关
function usePagination () {
  
  let paginationOptions = ref({
    page: 1,
    size: 10
  });

  const paginationOptionsChange = (newPaginationOptions: PaginationOptions) => {
    paginationOptions.value = reactive(newPaginationOptions);
  }

  return {
    paginationOptions,
    paginationOptionsChange
  }

}






