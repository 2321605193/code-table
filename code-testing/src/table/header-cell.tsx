

import { defineComponent } from '@vue/composition-api'

export default defineComponent({
  name: 'HeaderCell',
  setup(props, {slots}) {

    return () => {
      return (
        <th class='fj-table__header common-padding'>{slots?.default?.()}</th>
      )
    }
  },
})

