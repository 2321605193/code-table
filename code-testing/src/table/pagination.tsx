

import { defineComponent, reactive, inject, ref, computed, watch,  } from '@vue/composition-api'

export default defineComponent({
  name: 'Pagination',
  setup() {

    let sizeList = reactive([10,25,50,100])
    let { total, getSize, getPage, pageChange, sizeChange, } = inject('tableProvide')
    let paginationSize = ref(getSize.value);
    let paginationPage = ref(getPage.value);


    let pageList = computed(() => {
      return Math.ceil(total.value / paginationSize.value)
    })

    let paginationSizeChange = (e: Event) => {
      paginationSize.value = Number(e.target?.value) || 10;
      
    }

    let paginationPageChange = (e: Event) => {
      paginationPage.value = Number(e.target?.value) || 10;
      
    }

    watch(paginationSize, (newValue) => {
      if (pageList.value < paginationPage.value) {
        paginationPage.value = 1
      }
      sizeChange(newValue)
    })

    watch(paginationPage, (newValue) => {
      pageChange(newValue)
    })




    return () => {
      let sizeSelectOptions = sizeList.map((item) => {
        return (<option value={item}>{item}</option>);
      })

      let pageSelectOptions = []
      for (let index = 1; index <= pageList.value; index++) {

        pageSelectOptions.push(
          (<option value = {index}> {index} </option>)
        )
        
      }
      
      return (
        <section class='pagination-content'>
          <div>
            <span> 共{total.value}条 </span>
          </div>
          <div>
            <span>第</span>
            <select name="pageSelect" id="page" onChange={paginationPageChange}>
              {pageSelectOptions}
            </select>
            <span>页</span>
          </div>

          <div>
            <span>每页</span>
            <select  name="sizeSelect" id="size" onChange={paginationSizeChange}>
              {sizeSelectOptions}
            </select>
            <span>条</span>
          </div>
          

        </section>
      )
    }
  },


})

