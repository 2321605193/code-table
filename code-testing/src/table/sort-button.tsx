
import { computed, defineComponent, inject, ref, watch } from '@vue/composition-api'
import { isFunction } from '@vue/shared';
import ArrowDown from './arrow-down';


let defaultOrder = ['desc', 'asc', 'default'];
let defaultOrderBy = 'default'
export default defineComponent({
  props: {
    sortable: {
      type: Object
    },
    dataKey: {
      type: String,
      default: ''
    }
  },
  setup(props) {
    let  { changeSortRuleAndSortKey, getSortKey } = inject('tableProvide')
    let key = props.dataKey;

    let deafultSort = (curr: Record<string, any>, next: Record<string, any>) => curr[key] - next[key];
    let activeOrderBy = ref(defaultOrderBy)

    watch(() => getSortKey.value, () => {
      if (getSortKey.value !== key) {
        activeOrderBy.value = defaultOrderBy
      }
    })
    

   
    return () => {
      let sortFunction = (props.sortable?.sorter && isFunction(props.sortable?.sorter)) ? props.sortable.sorter : deafultSort
      let order = props.sortable?.orderBy ? props.sortable.orderBy : defaultOrder
    
      let oderFlag = computed(() => {
        return activeOrderBy.value === 'desc' ? -1 : 1
      })

      let sortData = () => {
        let nextOrderBy = order.indexOf(activeOrderBy.value) === -1 ? defaultOrder[0] : order[order.indexOf(activeOrderBy.value) + 1]
        activeOrderBy.value =  nextOrderBy ? nextOrderBy : defaultOrder[0]
        let sortRule = activeOrderBy.value === defaultOrderBy ? null : (curr: Record<string, any>, next: Record<string, any>) => oderFlag.value * sortFunction(curr, next);
        changeSortRuleAndSortKey(sortRule, key)
      }

      
      return (
          <span onClick={changeSortRuleAndSortKey(sortable, key)} class={[`sort-button sort-button__icon--${activeOrderBy.value}`]}>
            <i><ArrowDown /></i>
          </span>
      )
    }
  },
})
