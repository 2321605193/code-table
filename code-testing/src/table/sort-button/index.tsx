
import { defineComponent, inject, ref, watch } from '@vue/composition-api'
import ArrowDown from './arrow-down';
import {SortOrderBy} from '../types';




export default defineComponent({
  name: 'SortButton',
  props: {
    sortable: {
      type: [Object, Boolean],
      default: false
    },
    dataKey: {
      type: String,
      default: ''
    }
  },
  setup(props) {
    let  { setSortOptions, sortOptionsValue } = inject('tableProvide')
    let activeOrderBy = ref(SortOrderBy.default)

    watch([() => sortOptionsValue.value.sortKey, () => sortOptionsValue.value.activeOrderBy], () => {
      
      // 不是当前key置为默认
      if (sortOptionsValue.value.sortKey !== props.dataKey) {
        activeOrderBy.value = SortOrderBy.default
      } else {
        activeOrderBy.value = sortOptionsValue.value.activeOrderBy
      }
    })

    // 切换排序规则
    let sortData = () => {
      setSortOptions(props.sortable, props.dataKey)
    }
    
    return () => {
      return (
          <span onClick={sortData} class={[`sort-button sort-button__icon--${activeOrderBy.value}`]}>
            <i><ArrowDown /></i>
          </span>
      )
    }
  },
})
