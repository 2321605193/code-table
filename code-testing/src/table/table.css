.common-padding {
    padding: 10px 20px
}

.fj-table {
    display: inline-block;
    margin: auto;
    margin-top: 20px;
    margin-left: 100px;
    
    border: 1px solid #f0f0f0;
}

.fj-table table{
    width: 100%;
    border-collapse: collapse;
 }

.fj-table__row {
    border-bottom: 1px solid #f0f0f0;
}

