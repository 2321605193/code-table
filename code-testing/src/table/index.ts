/**
 * Created by uedc on 2021/10/11.
 */

import TestTable from './fj-table'
import './css/table.css'

export { TestTable }
