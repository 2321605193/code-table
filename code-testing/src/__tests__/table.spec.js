import { h, mount } from '@vue/test-utils'
import { TestTable } from '../table'

describe('Table', () => {
  const TableMount = options => mount(TestTable, options)

  test('render', () => {
    const wrapper = TableMount()
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  test('props showPagination test', () => {
    const wrapper = TableMount({
      propsData: {
        showPagination: true,
      },
    })
    expect(wrapper.find('.pagination-content').exists()).toBe(true)
  })

  test('props showHeader test', () => {
    const wrapper = TableMount({
      propsData: {
        showHeader: false,
      },
    })
    expect(wrapper.find('.fj-table__header').exists()).toBe(false)
  })


  test('props columns test', () => {
    const columns = [{
      title: '姓名',
      key: 'name',
    }, {
      title: '年龄',
      key: 'age',
      sortable: {
        orderBy: ['desc', 'asc'],
        sorter: (curr, next) => curr.age - next.age
      },
    }, {
      title: '爱好',
      key: 'love',
      render: (rowData) => {
        return 'love: ' + rowData.love
      },
      sortable: {
        orderBy: ['desc', 'asc', 'default'],
        sorter: (curr, next) => curr.love.length - next.love.length
      }
    }]
    const data = Array.from({ length: 100 }, (value, index) => {
      return {
        name: 'junjun' + index,
        age: 20 + index,
        love: Math.random().toString().substr(2, index / Math.floor(Math.random()*10)),
      }
    })
    const wrapper = TableMount({
      propsData: {
        data: data,
        columns: columns,
        showHeader: true,
        showIndex: true,
        showPagination: true
      },
    })

    expect(wrapper.find('.sort-button').exists()).toBe(true)
  })

})
