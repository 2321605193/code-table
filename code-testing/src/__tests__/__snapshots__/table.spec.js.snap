// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<section class=\\"fj-table\\">
  <section class=\\"fj-table-content\\">
    <table cellspacing=\\"0\\" cellpadding=\\"0\\" class=\\"fj-table-body\\">
      <thead>
        <tr class=\\"fj-table__row\\"></tr>
      </thead>
      <tbody></tbody>
    </table>
  </section><br>
  <section class=\\"pagination-content\\">
    <div><span> 共0条 </span></div>
    <div><span>第</span><select name=\\"pageSelect\\" id=\\"page\\"></select><span>页</span></div>
    <div><span>每页</span><select name=\\"sizeSelect\\" id=\\"size\\">
        <option value=\\"10\\">10</option>
        <option value=\\"25\\">25</option>
        <option value=\\"50\\">50</option>
        <option value=\\"100\\">100</option>
      </select><span>条</span></div>
  </section>
</section>"
`;
